core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -215,14 +227,110 @@ public boolean isAvailable() {
+
+      // Attempt to read the next entry in the file
+      pos += actualSize;
+      buf = readChannel(buf, pos, KEY_POS_LATEST);
+
+      // Only single FileEntry present with correct size, so data is not corrupt
+      if (buf.remaining() > 0)
+         return false;
+
+      buf.flip();
+      fe = new FileEntry(pos, buf);
+      channel.position(filePos);
+      return fe.size != fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen + 28 + (fe.expiryTime != -1 ? TIMESTAMP_BYTES : 0);
+   }
+
+   private void migrateCorruptData() throws IOException {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      long oldFilePos = MAGIC_LATEST.length;
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            oldFilePos += KEY_POS_LATEST;
+            // Explicitly calculate the correct size as incorrect value stored in fe.size
+            // Meta length incorrectly added 16 bytes, regardless of whether fe.expiryTime > 0
+            int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+
+            // Read old entry content and then write
+            ByRef.Long offset = new ByRef.Long(oldFilePos);
+
+            buf = readChannelUpdateOffset(buf, offset, fe.keyLen);
+            org.infinispan.commons.io.ByteBuffer key = readIntoByteBuffer(buf);
+
+            buf = readChannelUpdateOffset(buf, offset, fe.dataLen);
+            org.infinispan.commons.io.ByteBuffer value = readIntoByteBuffer(buf);
+
+            buf = readChannelUpdateOffset(buf, offset, metaLen);
+            org.infinispan.commons.io.ByteBuffer metadata = readIntoByteBuffer(buf);
+
+            // internalMetadata was ALWAYS lost - so don't read it
+
+            oldFilePos = offset.get();
+
+            // LastUsed and Created timestamps were lost, set creation time to current time if fe.expiryTime has been set
+            long currentTS = fe.expiryTime > 0 ? timeService.wallClockTime() : -1;
+            MarshallableEntry<? extends K, ? extends V> me = (MarshallableEntry<? extends K, ? extends V>) ctx.getMarshallableEntryFactory().create(key, value, metadata, null, currentTS, currentTS);
+            write(me, newChannel);
+
+            int corruptEntrySize = fe.keyLen + fe.dataLen + fe.metadataLen;
+            oldFilePos += corruptEntrySize;
```

Comment:
why do we need to update `oldFilePos` here? `readChannelUpdateOffset()` method should have updated the `offset` and I'm failing to understand why this line is required ...

Reply:
This line is required to increase the offset by the calculated `corruptEntrySize` bytes so that the erroneous bits that were written during the corrupt migration are skipped before attempting to read the next `FileEntry`

Reply:
I need to attach a debugger because I'm not understanding how is this working as it is.
This is what I see
```
line 281: offset = oldFilePos
line 283: offset = offset + fe.keyLen
line 286: offset = offset + fe.dataLen
line 289: offset = offset + metaLen
line 294: oldFilePos = offset // which is the same as oldFilePos = oldFilePos + fe.keyLen + fe.dataLen + metaLen
line 302: oldFilePos = oldFilePos + fe.keyLen + fe.dataLen + fe.metadataLen // again?
```

Reply:
```
line 302: oldFilePos = oldFilePos + fe.keyLen + fe.dataLen + fe.metadataLen // again?
```

This is simulating how the old `migrateFromV11` incorrectly incrementing the  `newFilePos` var https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java#L332-L345

Reply:
eventually, I got there :)
in the corrupted file I was excepting to see the key duplicated but it just zeros